### ISSUE TITLE
terminal should be login shell, not interactive shell

### ISSUE THREAD

**Issue by bklaas** [jupyterlab/jupyterlab#3094]
The summary of this issue is that I think when initially launching a terminal it should start a login shell, not a shell in interactive mode.

I have a jupyterlab instance running via jupyterhub on a shared (ubuntu) linux server. Default shell for all users is bash. 

When I launch a terminal in jupyterlab, it sources my .bashrc file, which in turn sets my PS1 prompt with some dynamic git information via the __git_ps1 bash function that the git package provides and is made available in login shells via /etc/bash_completion.d/git-prompt. The scripts in bash_completion.d/ are only sourced if you are in a login shell. As a result, every time I hit return in the terminal it throws a "__git_ps1: command not found" error.

We also have some environmental variables, e.g. PERL5LIB, that are set inside /etc/profile.d/ scripts. Again, these are not sourced in jupyterhub terminal because it's not a login shell.

We've worked around this by setting PERL5LIB in /etc/environment (not a great solution from a devops perspective, but works in interactive shells) and hacking my .bashrc to source /etc/bash_completion.d/git-prompt.

Nevertheless, it seems reasonable to think that a terminal in jupyterhub should begin life as a login shell.

**Comment by blink1073** [comment]
@takluyver, thoughts on this?  

The default shell command is handled [here](https://github.com/jupyter/notebook/blob/a8be6598b9b4e30ca88cdc9a437e3541c65a624b/notebook/terminal/__init__.py#L16). 

It is configurable by setting config for `c.NotebookApp.terminado_settings = { "shell_command": "foo" }`


https://github.com/jupyter/notebook/blob/master/notebook/notebookapp.py#L770
http://jupyter-notebook.readthedocs.io/en/stable/config_overview.html#configure-common

**Comment by blink1073** [comment]
cc @minrk

**Comment by takluyver** [comment]
I'd expect it to do the same as launching a terminal emulator inside my desktop, which runs `.bashrc` but not `.bash_profile` (i.e. it is an interactive shell, not a login shell).

I guess there's an argument that Jupyterhub should start it's singleuser server through a login shell: that is the step where you become a logged in user, analogous to logging in to your graphical desktop. 

Running a login shell for a graphical login has been a point of controversy with the move to Wayland: initially it didn't run through a login shell, and there was talk of a more declarative way to set environment variables on login. But it wasn't happening quickly enough, so they relented and made it run a login shell again.

**Comment by minrk** [comment]
Arg, it's always annoyed me that login and interactive shells are different.

@takluyver is right that to get probably the most expected behavior is to launch the notebook server via a login shell. One way to do that is to make a script that itself tells bash to invoke a login shell, and then launch jupyterhub-singleuser:

```bash
#!/usr/bin/env bash -l

# any additional initialization can go here, too
export MYVAR=foo
source /etc/jupyter/extra-bash.d/...

# launch jupyter (exec and $@ are important!)
exec jupyterhub-singleuser $@
```

and then change `c.Spawner.cmd = [ '/path/to/my/script.sh' ]`.

**Comment by bubthegreat** [comment]
I can see use cases for both, but it doesn't look like this is a configurable option via the jupyterhub interface.  I'd personally like this to be something I can configure in the settings -> Terminal settings -> default shell

**Comment by bklaas** [comment]
A polite bump to ask this gets included in a 1.0 release.

Respectful of the previous comment, however I'm not sure there is a real use case for a new terminal inside jupyterlab not being a login shell. To be clear, I don't mean requiring the user to manually login to every new terminal shell, I mean sourcing the startup files that are sourced in a login shell context.

**Comment by Monduiz** [comment]
I have to add my voice to this as an important thing to add. We are using jupyterhub with Jupiter Lab to coordinate work on virtual machines and it simplified our workflows by a lot. However, if we need to addd something to our condo environments, we have to ssh into the VMs because the Jupiter Lab terminal is not functional as it's not sourcing from login configurations.

**Comment by jasongrout** [comment]
From the conversation above, it sounds like this is an issue in the notebook server, not JupyterLab. Do you have the same issue in the terminal launched by the classic notebook?

**Comment by bklaas** [comment]
Huh, you learn something every day, I wasn't even aware the terminal was delivered via the Notebook server :)

I can confirm that the same issue is seen with the terminal launched by the classic notebook.

Can you give a pointer as to how best to push this issue towards the right project? Still very interested in this issue getting some attention.

**Comment by Monduiz** [comment]
Right, I should have thought of that too. We push this to jupyterhub or jupyter.

**Comment by jasongrout** [comment]
@blink1073's [comment above](https://github.com/jupyterlab/jupyterlab/issues/3094#issuecomment-336572126) points out the code location for the default terminal command in the https://github.com/jupyter/notebook/ repo. I think that repo is probably the appropriate place for discussing this change.

**Comment by jasongrout** [comment]
(Or since it's a setting you can set, perhaps whoever launches the notebook should be responsible for something like this, i.e., jupyterhub?)

**Comment by Monduiz** [comment]
@jasongrout I somehow missed that. If it can bet changed using the config file, it is a simple matter.

**Comment by jamesdbrock** [comment]
It looks like the shell was forced to be a login shell in this commit from Oct 18, 2018

https://github.com/jupyter/notebook/commit/f859abde690e60f778bfeb0f5ee08bfbfa9fef12

This was done by appending the `-l` switch to the `shell_command`, so now it seems that it's impossible to make the login shell interactive by setting the `shell_command` like:

    c.NotebookApp.terminado_settings={'shell_command':['bash', '-i']}

because the `-l` switch will come after and override the `-i` switch.

**Comment by jasongrout** [comment]
Interesting. Again, it sounds more appropriate to bring this up in the https://github.com/jupyter/notebook/ repo, where that code is.

**Comment by bklaas** [comment]
I'm going to mark this issue as closed because it's not a jupyterlab issue

As a closing point to anyone reading this issue though, the checkin to notebook made Oct 18, 2018 referenced above still hasn't been pushed out in an actual release. 

The fix is merged into master, slated to be included with the 5.8 release of jupyter/notebook
